go-controller/pkg/network-attach-def-controller/network_attach_def_controller_test.go

```diff
@@ -399,27 +459,37 @@ func TestNetAttachDefinitionController(t *testing.T) {
 			}
 
 			meetsExpectations := func(g gomega.Gomega) {
-				tncm.Lock()
-				defer tncm.Unlock()
-
 				var expectRunning []string
 				for _, expected := range tt.expected {
 					netInfo, err := util.NewNetInfo(expected.network)
 					g.Expect(err).ToNot(gomega.HaveOccurred())
 
 					name := netInfo.GetNetworkName()
 					testNetworkKey := testNetworkKey(netInfo)
-
-					// test that the controller have the expected config and NADs
-					g.Expect(tncm.controllers).To(gomega.HaveKey(testNetworkKey))
-					g.Expect(tncm.controllers[testNetworkKey].Equals(netInfo)).To(gomega.BeTrue(),
-						fmt.Sprintf("matching network config for network %s", name))
-					g.Expect(tncm.controllers[testNetworkKey].GetNADs()).To(gomega.ConsistOf(expected.nads),
-						fmt.Sprintf("matching NADs for network %s", name))
+					func() {
+						tncm.Lock()
+						defer tncm.Unlock()
```

Comment:
Just wondering why you can't hold this lock throughout `meetsExpectations`

Reply:
it caused deadlock, because normal locking order is:
1. lock NAD controller
2. lock NM
3. tncm lock

When I introduced GetActiveNetworkNamespace to the test, the lock order would be:
1. lock tncm
2. lock NAD controller
3. lock NM

That can cause deadlock